Ls 186 scaling cluster (#32)
datastax/pulsar-docs: modules/ROOT/pages/scale-cluster.adoc

----
$ helm install pulsar -f ~/dev-values_large.yaml --wait datastax-pulsar/pulsar
----

== Generate data

. Generate data in our namespaces:

----
$ bin/pulsar-admin namespaces set-retention public/default -t -1 -s 100M
$ bin/pulsar-client produce test1 -n 500 -m "sdasdsdsadsadsad"
$ bin/pulsar-client produce test2 -n 500 -m "sdasdsdsadsadsad"
$ bin/pulsar-client produce test3 -n 500 -m "sdasdsdsadsadsad"
----
[start=2]
. Ensure all bookies have some ledgers assigned to them:

----
$ ./bin/bookkeeper shell listledgers -meta | grep pulsar-bookkeeper-3
----

== Destroy Persistent Volume Claim (PVC)

To downscale a bookie and destroy its persistent volume claim (PVC):

----
 bookkeeper:
    replicaCount: 3

$ helm upgrade pulsar -f ~/dev-values_large.yaml --wait datastax-pulsar/pulsar
----

== Delete PVC

To delete the PVC:
----
$ kubectl delete pvc pulsar-bookkeeper-ledgers-pulsar-bookkeeper-3
----

== Scale Back Cluster
To scale back the cluster:
----
 bookkeeper:
    replicaCount: 4

$ helm upgrade pulsar -f ~/dev-values_large.yaml --wait datastax-pulsar/pulsar
----

Now, bookkeeper-3 will fail to start with
`org.apache.bookkeeper.bookie.BookieException$InvalidCookieException`.

> == Recovery from lost PVC
> 
> . Track under-replicated ledgers from any live bookie:
> ----
> $ ./bin/bookkeeper shell listunderreplicated
> ----
> [start=2]
> . Double-check bookie id of the failing bookie:
> ----
> $ ./bin/bookkeeper shell listbookies -a
> ----
> 
> This should return `Cannot resolve pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181, bookie is unknown`.
> [start=3]
> . Decommission the bookie:
> 
> ----
> $ ./bin/bookkeeper shell decommissionbookie -bookieid pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181
> ----
> [start=4]
> . Cookie of the decommissioned bookie: `pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181` is deleted successfully.
> . Restart failing bookie with Ctrl-K in k9s, or wait for k8s to restart it again.
> . The bookie should come up and rejoin the cluster (data is lost, as expected with deleted PVC).

Reviewer: ```suggestion
== Recovering from lost PVC

. Track under-replicated ledgers from any live bookie:
+
[source,shell]
----
$ ./bin/bookkeeper shell listunderreplicated
----

. Double-check bookie id of the failing bookie:
+
[source,shell]
----
$ ./bin/bookkeeper shell listbookies -a
----
+
This command returns `Cannot resolve pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181, bookie is unknown`.

[start=3]
. Decommission the bookie:
+
[source,shell]
----
$ ./bin/bookkeeper shell decommissionbookie -bookieid pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181
----
+
The cookie of the decommissioned bookie: `pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181` is deleted.

. Restart failing bookie with Ctrl-K in k8s or wait for k8s to restart it again.
+
The bookie should restart and rejoin the cluster. As expected with deleted PVC, data is lost.
```

== Better scale down procedure

1. Ensure there are no under replicated ledgers
2. Downscale the cluster. New under replicated ledgers will appear.

> 3. Decommission the removed bookie (bookie has to be down). This will take some time because it triggers the audit and autorecovery, and blocks until autorecovery completes. 

Reviewer: What is blocked until the auto-recovery is completed?

> 4. There is an option to scale back if something fails

Reviewer: How do they scale back if needed?

> 1. Ensure there are no under replicated ledgers
> 2. Downscale the cluster. New under replicated ledgers will appear.
> 3. Decommission the removed bookie (bookie has to be down). This will take some time because it triggers the audit and autorecovery, and blocks until autorecovery completes. 
> 4. There is an option to scale back if something fails
> 5. Delete the PVCs of decommissioned bookie (both ledger and journal).

Reviewer: ```suggestion
. Ensure there are no under-replicated ledgers.
. Downscale the cluster.
+ 
New under replicated ledgers will appear.
. Decommission the removed bookie.
+
[NOTE]
====
The decommissioning will take some time because it triggers the audit and auto-recovery and blocks until auto-recovery completes. 
====
+
[NOTE]
====
There is an option to scale back if something fails.
====

. Delete the ledger and journal PVCs of decommissioned bookie.
```

> 5. Delete the PVCs of decommissioned bookie (both ledger and journal).

Reviewer: Do ledger and journal describe PVC or something else?

>     defaultWriteQuorum: 3
----

. Create the cluster by installing Pulsar with `dev-values_large.yaml`:
+
----
$ helm install pulsar -f ~/dev-values_large.yaml --wait datastax-pulsar/pulsar
----

== Generating data

. Generate data in your namespaces:
+
----
$ bin/pulsar-admin namespaces set-retention public/default -t -1 -s 100M
$ bin/pulsar-client produce test1 -n 500 -m "sdasdsdsadsadsad"
$ bin/pulsar-client produce test2 -n 500 -m "sdasdsdsadsadsad"
$ bin/pulsar-client produce test3 -n 500 -m "sdasdsdsadsadsad"
----

. Ensure all bookies have some ledgers assigned to them:
+
----
$ ./bin/bookkeeper shell listledgers -meta | grep pulsar-bookkeeper-3
----

== Downscaling bookies

. To downscale a bookie, change the `replicaCount` value in the YAML file to a number less than the current `replicaCount`:
+
----
 bookkeeper:
    replicaCount: 3
----
+
. Upgrade the Helm chart to use the new value:
+
----
$ helm upgrade pulsar -f ~/dev-values_large.yaml --wait datastax-pulsar/pulsar
----
+
Now, `bookkeeper-3` will fail to start with `org.apache.bookkeeper.bookie.BookieException$InvalidCookieException`.

> +

Reviewer: ```suggestion

```

. Destroy the bookie's PVC:
----
$ kubectl delete pvc pulsar-bookkeeper-ledgers-pulsar-bookkeeper-3
----

== Scaling up your cluster

To scale up your cluster, change the `replicaCount` value in the YAML file to a number greater than the current `replicaCount`:

[source,shell]
----
 bookkeeper:
    replicaCount: 5
----

> +

Reviewer: ```suggestion

```

. Upgrade the Helm chart to use the new value in the Pulsar cluster:
+
----
$ helm upgrade pulsar -f ~/dev-values_large.yaml --wait datastax-pulsar/pulsar
----

== Recovering from a lost Persistent Volume Claim (PVC)

. If a PVC goes down, track under-replicated ledgers from any live bookie:
+
[source,shell]
----
$ ./bin/bookkeeper shell listunderreplicated
----

. Double-check the bookie id of the failing bookie:
+
[source,shell]
----
$ ./bin/bookkeeper shell listbookies -a
----
+
This command will return `Cannot resolve pulsar-bookkeeper-3.pulsar-bookkeeper.default.svc.cluster.local:3181, bookie is unknown`.

> +

Reviewer: ```suggestion

```
